README.md
# Microlock
A dead simple distributed locking library for Node.js and [etcd](http://github.com/coreos/etcd) (via [node-etcd](https://github.com/stianeikeland/node-etcd))

[![NPM](https://nodei.co/npm/microlock.png)](https://nodei.co/npm/microlock/)

[![CircleCI](https://circleci.com/gh/thebigredgeek/microlock.svg?style=shield)](https://circleci.com/gh/thebigredgeek/microlock/tree/master)




## What is Etcd?

[Etcd](https://github.com/coreos/etcd) is a distributed key-value store, built by the [CoreOS](https://coreos.com/) team, that provides strong guarantees around consistency and partition tolerance.  Data is duplicated to all nodes in a given cluster and remains consistent between node failures.  Cluster leaders are elected via the [Raft consensus algorithm](https://raft.github.io/).  Etcd provides operations for atomic value swapping/removal based on criteria and TTL for values, making it a perfect media for distributed locks.

## What is a distributed lock?

A distributed lock is a mechanism that provides serialized flow control on a context that is acted on by more than one process.  These processes typically operate on different machines via Service Oriented Architecture.  Each process uses an object called a distributed lock to "lock" access to the shared context, aliased by a key, so that only one process, each aliased by a node id, can act on it at a time, thereby ensuring consistency and preventing race conditions.

## Why not Redlock?

Redis is great for a lot of things.  Caching, keeping processes stateless, and fast access to simply structured data are all cases where Redis shines.  However, implementing a distributed lock with Redis via Redlock has several caveats that are unsuitable for many cases.  Namely, if you need strong guarantees that a lock will not be acquired by multiple nodes at once even in the event of failure, Redlock isn't a viable option.

## Notes

Microlock is currently compatible with Etd 2.2.x.  Work on support for Etcd 2.2.x - 3.x is in progress.

## Install
**Requires NodeJS >= 4.0**

```bash
$ npm install microlock
```

## Basic usage

### ES5
```javascript
var os = require('os');
var Etcd = require('node-etcd');
var Microlock = require('microlock');

var key = 'foo'; //name of the lock
var id = os.hostname(); //id of *this* node
var ttl = 5; //5 second lease on lock

var etcd = new Etcd();
var foo = new Microlock.default(etcd, key, id, ttl);

foo.lock().then(function () {
  // foo is locked by this node

  // do some stuff...

  // release the lock
  return foo.unlock();
}, function (e) {
  if (e instanceof Microlock.AlreadyLockedError) {
    // foo is already locked by a different node
  }
});
```

### ES2015 (with babel)
```javascript
import { hostname } from 'os';
import Etcd from 'node-etcd';
import Microlock, { AlreadyLockedError } from 'microlock';

const key = 'foo'; //name of the lock
const id = hostname(); //id of *this* node
const ttl = 5; //5 second lease on lock

const etcd = new Etcd();
const foo = new Microlock(etcd, key, id, ttl);

foo.lock().then(() => {
  // foo is locked by this node

  // do some stuff...

  // release the lock
  return foo.unlock();
}, (e) => {
  if (e instanceof AlreadyLockedError) {
    // foo is already locked by a different node
  }
});
```

### ES2016/2017 (with babel)
```javascript

import { hostname } from 'os';
import Etcd from 'node-etcd';
import Microlock, { AlreadyLockedError } from 'microlock';

async function main () {

  const key = 'foo'; //name of the lock
  const id = hostname(); //id of *this* node
  const ttl = 5; //5 second lease on lock

  const etcd = new Etcd();
  const foo = new Microlock(etcd, key, id, ttl);

  try {
    await foo.lock();
    // foo is locked by this node

    // do some stuff...

    // release the lock
    await foo.unlock();
  } catch (e) {
    if (e instanceof AlreadyLockedError) {
      // foo is already locked by a different node
    }
  }
}

main();
```

## Methods

All methods (except destroy) return promises, making it easy to use features like async/await with ES2016/ES2017 via Babel.

### Microlock(etcd, key, node_id, [ttl = 1])
Creates a microlock client for a lock key.

```javascript
var Etcd = require('node-etcd');
var Microlock = require('microlock');

var etcd = new Etcd();
var foo = new Microlock.default(microlock, 'foo', 'bar');
```

### .lock()
Attempts to lock the `key` for the `node_id`.

```javascript
foo.lock().then(function () {
  // foo is locked by this node
}, function (e) {
  if (e instanceof Microlock.AlreadyLockedError) {
    // foo is already locked by a different node
  }
});
```

### .unlock()
Attempts to release the `key` for the `node_id`.

```javascript
foo.unlock().then(function () {
  // foo is unlocked
}, function (e) {
  if (e instanceof Microlock.LockNotOwnedError) {
    // foo is not locked by `node_id`
  }
})
```

### .renew()
Attempts to renew the lock on `key` for the `node_id`.

```javascript
foo.renew().then(function () {
  // foo lease is renewed... ttl is refreshed
}, function (e) {
  if (e instanceof Microlock.LockNotOwnedError) {
    // foo is not locked by `node_id`
  }
})
```

### .destroy()
Unbinds listeners/watchers from this client

## Events

### unlock
Emits when the key is unlocked (node agnostic)

```javascript
foo.on(Microlock.events.unlocked, function () {
  //handle unlocked with constant
});

foo.on('unlocked', function () {
  //handle unlocked with string
});
```

### locked
Emits when the key is locked (node agnostic)

```javascript
foo.on(Microlock.events.locked, function () {
  //handle locked with constant
});

foo.on('locked', function () {
  //handle locked with string
});
```

## Contributing

### Installing packages

```bash
  $ make install
```

### Building

```bash
  $ make
```

### Linting

```bash
  $ make lint
```

### Running unit tests

```bash
  $ make unit
```

### Running integration tests

**[Docker Compose](https://docs.docker.com/compose/) is required**

```bash
  $ make integration etcd_image_version=v2.2.2
```

You can use whatever version you'd like to test against in the command above.


CHANGELOG.md
1.0.2 / 2016-11-10
==================

* Stable build

1.2.1 / 2018-08-15
==================

* Send null for the value during lock renewal (#6)
* Don't throw lock-not-owned when that isn't the error (#6)


test/setup.js
//require('source-map-support').install();
require('babel-register');
require('babel-polyfill');

function kill (e) {
  console.trace(e);
  process.exit(1);
}

//process.on('uncaughtException', kill);
//process.on('unhandledRejection', kill);


test/unit/microlock_spec.js
import EventEmitter from 'events';
import { expect } from 'chai';
import Promise from 'bluebird';
import sinon from 'sinon';

import Microlock, {
  events,

  EtcdClientRequiredError,
  KeyRequiredError,
  NodeIdRequiredError,
  InvalidTtlError,
  AlreadyLockedError,
  LockNotOwnedError
} from '../../src/microlock';

class MockWatcher extends EventEmitter {
  constructor () {
    super();

    sinon.stub(this, 'on', this.on);
    sinon.stub(this, 'stop', this.close);
  }
  stop () {
    this.removeAllListeners();
  }
}

class MockEtcdClient {
  constructor (opts = {}) {
    const { setErr, setRes, compareAndDeleteErr, compareAndDeleteRes, watcher } = opts

    this.__setErr = setErr;
    this.__setRes = setRes;
    this.__compareAndDeleteErr = compareAndDeleteErr;
    this.__compareAndDeleteRes = compareAndDeleteRes;
    this.__watcher = watcher;

    sinon.stub(this, 'watcher', this.watcher)
    sinon.stub(this, 'set', this.set);
    sinon.stub(this, 'compareAndDelete', this.compareAndDelete);
  }
  watcher () {
    return this.__watcher
  }
  set (key, val, opts, cb) {
    if (this.__setErr) return cb(this.__setErr, null);
    return cb(null, this.__setRes);
  }
  compareAndDelete (key, val, cb) {
    if (this.__compareAndDeleteErr) return cb(this.__compareAndDeleteErr, null);
    return cb(null, this.__compareAndDeleteRes);
  }
}

describe('microlock', () => {
  let watcher = null
    , setRes = null
    , compareAndDeleteRes = null
    , etcd = null;
  beforeEach(() => {
    watcher = new MockWatcher();
    setRes = 'foo';
    compareAndDeleteRes = 'bar';
    etcd = new MockEtcdClient({
      watcher,
      setRes,
      compareAndDeleteRes
    });
  })
  context('etcd client not provided', () => {
    it('throws EtcdClientRequiredError', () => {
      try {
        new Microlock(null);
        throw new Error('should have thrown EtcdClientRequiredError');
      } catch (e) {
        expect(e instanceof EtcdClientRequiredError).to.be.true;
      }
    });
  });
  context('key is not a string', () => {
    it('throws KeyRequiredError', () => {
      try {
        new Microlock(watcher, 123);
        throw new Error('should have thrown KeyRequiredError');
      } catch (e) {
        expect(e instanceof KeyRequiredError).to.be.true;
      }
    });
  });
  context('node_id is not a string', () => {
    it('throws NodeIdRequiredError', () => {
      try {
        new Microlock(watcher, 'foo', 123);
        throw new Error('should have thrown NodeIdRequiredError');
      } catch (e) {
        expect(e instanceof NodeIdRequiredError).to.be.true;
      }
    });
  });
  context('ttl is not a number', () => {
    it('throws InvalidTtlError', () => {
      try {
        new Microlock(watcher, 'foo', 'bar', 'hello');
        throw new Error('should have thrown InvalidTtlError');
      } catch (e) {
        expect(e instanceof InvalidTtlError).to.be.true;
      }
    });
  });
  describe('watcher', () => {
    let microlock = null;
    beforeEach(() => {
      microlock = new Microlock(etcd, 'foo', 'bar');
    })
    context('on action "compareAndDelete"', () => {
      it('emits events.unlocked', (done) => {
        microlock.on(events.unlocked, done);
        watcher.emit('compareAndDelete');
      });
    });
    context('on action "change"', () => {
      context('if node contains ttl', () => {
        it('emits events.locked', (done) => {
          microlock.on(events.locked, done);
          watcher.emit('change', {
            node: {
              ttl: 123
            }
          });
        });
      });
    });
  });
  describe('method', () => {
    let key = null
      , node_id = null
      , ttl = null
      , microlock = null;
    beforeEach(() => {
      key = 'foo';
      node_id = 'bar';
      ttl = 123;
      microlock = new Microlock(etcd, key, node_id, ttl);
    })
    describe('destroy', () => {
      it('stops watchers', () => {
        microlock.destroy();
        expect(watcher.stop.calledOnce).to.be.true;
      });
      it('removes all event listeners', () => {
        sinon.stub(microlock, 'removeAllListeners');
        microlock.destroy();
        expect(microlock.removeAllListeners.calledOnce).to.be.true;
      });
    });
    describe('lock', () => {
      context('when able to verify unlocked state', () => {
        it('sets the state to locked', () => {
          microlock.lock();
          expect(etcd.set.calledWith(key, node_id, {
            ttl,
            prevExist: false
          })).to.be.true;
        });
        it('returns a resolving promise passing the etcd response', (done) => {
          microlock.lock()
          .then((res) => expect(res).to.equal(setRes))
          .then(() => done(), done);
        });
      });
      context('when unable to verify locked', () => {
        beforeEach(() => {
          etcd.__setErr = new Error();
          etcd.__setErr.errorCode = 105;
        });
        it('returns a rejecting promise passing AlreadyLockedError', (done) => {
          microlock.lock()
          .catch((e) => expect(e instanceof AlreadyLockedError).to.be.true)
          .then(() => done(), done);
        });
      });
    });
    describe('unlock', () => {
      context('when able to verify locked state by current node', () => {
        it('sets the state to unlocked', () => {
          microlock.unlock();
          expect(etcd.compareAndDelete.calledWith(key, node_id)).to.be.true;
        });
        it('returns a resolving promise passing the etcd response', (done) => {
          microlock.unlock()
          .then((res) => expect(res).to.equal(compareAndDeleteRes))
          .then(() => done(), done)
        });
      });
      context('when not able to verify locked state by current node', () => {
        it('returns a rejecting promise passing LockNotOwnedError', (done) => {
          microlock.unlock()
          .catch((e) => expect(e instanceof LockNotOwnedError).to.be.true)
          .then(() => done(), done);
        });
      });
    });
    describe('renew', () => {
      context('when able to verify locked stat by current node', () => {
        it('sets the state to locked with a refreshed ttl', () => {
          microlock.renew();
          expect(etcd.set.calledWith(key, null, {
            ttl,
            prevValue: node_id,
            refresh: true
          })).to.be.true;
        });
        it('returns a resolving promise passing the etcd response', (done) => {
          microlock.renew()
          .then((res) => expect(res).to.equal(setRes))
          .then(() => done(), done);
        });
      });
      context('when not able to verify locked state by current node', () => {
        beforeEach(() => {
          etcd.__setErr = new Error();
          etcd.__setErr.errorCode = 101;
        });
        it('returns a rejecting promise passing LockNotOwnedError', (done) => {
          microlock.renew()
          .catch((e) => expect(e instanceof LockNotOwnedError).to.be.true)
          .then(() => done(), done);
        });
      });
    });
  });
});


test/integration/microlock_spec.js
import { expect } from 'chai';
import sinon from 'sinon';
import Promise from 'bluebird';
import Etcd from 'node-etcd';

import Microlock, {
  events,

  EtcdClientRequiredError,
  KeyRequiredError,
  NodeIdRequiredError,
  InvalidTtlError,
  AlreadyLockedError,
  LockNotOwnedError
} from '../../src/microlock';

describe('microlock', () => {
  let etcd = null
    , key = null
    , node1 = null
    , node2 = null
    , lock1 = null
    , lock2 = null
    , ttl = null;

  beforeEach(() => {
    etcd = new Etcd();

    key = 'key';

    node1 = 'node_1';
    node2 = 'node_2';

    ttl = 5;

    lock1 = new Microlock(etcd, key, node1, ttl);
    lock2 = new Microlock(etcd, key, node2, ttl);
  });

  afterEach(() => Promise.all([
    lock1.unlock(),
    lock2.unlock()
  ]).catch((e) => e));

  describe('lock', () => {
    context('is locked', () => {
      beforeEach(() => lock2.lock());
      it('rejects with AlreadyLockedError', (done) => {
        lock1
        .lock()
        .then(() => done(new Error('should have rejected')), (e) => {
          expect(e instanceof AlreadyLockedError).to.be.true;
          done();
        })
        .catch(done);
      });
    });
    context('is not locked', () => {
      it('resolves', (done) => {
        lock1
        .lock()
        .then(() => done(), () => done(new Error('should have resolved')));
      });
    });
  });
  describe('unlock', () => {
    context('is locked by current node', () => {
      beforeEach(() => lock1.lock());
      it('resolves', (done) => {
        lock1
        .unlock()
        .then(() => done(), () => done(new Error('should have resolved')));
      });
    });
    context('is not locked by current node', () => {
      beforeEach(() => lock2.lock());
      it('rejects with LockNotOwnedError', (done) => {
        lock1
        .unlock()
        .then(() => done(new Error('should have rejected')), (e) => {
          expect(e instanceof LockNotOwnedError).to.be.true;
          done();
        })
        .catch(done);
      });
    });
    context('is not locked', () => {
      it('rejects with LockNotOwnedError', (done) => {
        lock1
        .unlock()
        .then(() => done(new Error('should have rejected')), (e) => {
          expect(e instanceof LockNotOwnedError).to.be.true;
          done();
        })
        .catch(done);
      });
    });
  });
  describe('renew', () => {
    context('is locked by current node', () => {
      beforeEach(() => lock1.lock());
      it('resolves', (done) => {
        lock1
        .renew()
        .then(() => done(), () => done(new Error('should have resolved')));
      });
    });
    context('is not locked by current node', () => {
      beforeEach(() => lock2.lock());
      it('rejects with LockNotOwnedError', (done) => {
        lock1
        .renew()
        .then(() => done(new Error('should have rejected')), (e) => {
          expect(e instanceof LockNotOwnedError).to.be.true;
          done();
        })
        .catch(done);
      });
    });
    context('is not locked', () => {
      it('rejects with LockNotOwnedError', (done) => {
        lock1
        .renew()
        .then(() => done(new Error('should have rejected')), (e) => {
          expect(e instanceof LockNotOwnedError).to.be.true;
          done();
        })
        .catch(done);
      });
    });
  });
});


src/microlock.js
import EventEmitter from 'events';
import Promise from 'bluebird';



// Error helper
function ExtendableError(message){
  Error.apply(this, arguments);
  if (typeof Error.captureStackTrace === 'function') {
    Error.captureStackTrace(this, this.constructor);
  } else {
    this.stack = (new Error(message)).stack;
  }
}

ExtendableError.prototype = Object.create(Error.prototype);

Object.setPrototypeOf(ExtendableError, Error);







// Errors
export class EtcdClientRequiredError extends ExtendableError {
  constructor () {
    super('etcd client parameter is required');
  }
}

export class KeyRequiredError extends ExtendableError {
  constructor () {
    super('key parameter of type String is required');
  }
}

export class NodeIdRequiredError extends ExtendableError {
  constructor () {
    super('node_id parameter of type String is required')
  }
}

export class InvalidTtlError extends ExtendableError {
  constructor () {
    super('ttl parameter must be of type Number');
  }
}

export class AlreadyLockedError extends ExtendableError {
  constructor (key) {
    super(`Lock "${key}" is already locked`);
  }
}

export class LockNotOwnedError extends ExtendableError {
  constructor (key, node_id) {
    super(`Lock "${key}" is not owned by node "${node_id}"`);
  }
}








// Events
export const events = {
  locked: 'locked',
  unlocked: 'unlocked',
  error: 'error'
};







//Microlock
export default class Microlock extends EventEmitter {
  constructor (etcd, key, node_id, ttl = 1) {

    if (!etcd) throw new EtcdClientRequiredError();
    if (typeof key !== 'string') throw new KeyRequiredError();
    if (typeof node_id !== 'string') throw new NodeIdRequiredError();
    if (isNaN(ttl)) throw new InvalidTtlError();

    super();

    this.__etcd = etcd;
    this.__key = key;
    this.__node_id = node_id;
    this.__ttl = ttl < 1 ? 1 : ttl; //Require at least one second for ttl for stability of lock/unlock

    this.__watcher = this.__etcd.watcher(key);

    this.__watcher.on('compareAndDelete', () => this.emit(events.unlocked));
    this.__watcher.on('change', (v) => {
      if (v.node && v.node.ttl) this.emit(events.locked);
    });
  }

  // Private methods
  __throwAlreadyLocked () {
    const e = new AlreadyLockedError(this.__key);
    this.emit(events.error, e)
    throw e;
  }
  __throwLockNotOwned () {
    const e = new LockNotOwnedError(this.__key, this.__node_id);
    this.emit(events.error, e);
    throw e;
  }


  // Public methods
  destroy () {
    this.__watcher.stop();
    this.removeAllListeners();
  }

  lock () {
    return new Promise((resolve, reject) => {
      return this.__etcd.set(this.__key, this.__node_id, {
        ttl: this.__ttl,
        prevExist: false
      }, (err, val) => {
        if (err) return reject(err);
        return resolve(val)
      });
    })
    .catch((e) => {
      // Key already exists
      if (e.errorCode === 105) {
        this.__throwAlreadyLocked();
      }
      throw e;
    });
  }

  unlock () {
    return new Promise((resolve, reject) => {
      return this.__etcd.compareAndDelete(this.__key, this.__node_id, (err, res) => {
        if (err) return reject(err);
        return resolve(res);
      });
    })
    .catch((e) => {
      // KeyNotFound or CompareFailed
      if (e.errorCode === 100 || e.errorCode === 101) {
        this.__throwLockNotOwned();
      }
      throw e;
    });
  }

  renew () {
    return new Promise((resolve, reject) => {
      return this.__etcd.set(this.__key, null, {
        prevValue: this.__node_id,
        refresh: true,
        ttl: this.__ttl
      }, (err, val) => {
        if (err) return reject(err);
        return resolve(val);
      })
    })
    .catch((e) => {
      // KeyNotFound or CompareFailed
      if (e.errorCode === 100 || e.errorCode === 101) {
        this.__throwLockNotOwned();
      }
      throw e;
    });
  }
}
